editor/modes/visual-editor/block.js

```diff
@@ -51,13 +51,28 @@ import {
 	getMultiSelectedBlockUids,
 } from '../../selectors';
 
-const { BACKSPACE, ESCAPE, DELETE, UP, DOWN, LEFT, RIGHT, ENTER } = keycodes;
+const { BACKSPACE, ESCAPE, DELETE, UP, DOWN, LEFT, RIGHT, ENTER, TAB } = keycodes;
+const F10 = 121;
 
 function FirstChild( { children } ) {
 	const childrenArray = Children.toArray( children );
 	return childrenArray[ 0 ] || null;
 }
 
+function selectAncestor( node, selector ) {
+	while ( node !== null ) {
+		if ( node.querySelector( selector ) !== null ) {
+			return node;
+		}
+		node = node.parentNode;
+	}
+	return null;
+}
+
+function isToolbar( target ) {
+	return selectAncestor( target, '.components-toolbar, .editor-block-settings-menu, .editor-block-mover' ) !== null;
```

Comment:
It's a very tricky problem because the very purpose of what you're implementing is to manage the relationship between descendant components (i.e. moving from blocks into their toolbar). @youknowriad has been exploring some ideas in #2424 (#2421) that at least have an advantage of isolating these behaviors, if not eliminating the DOM querying.

This is also an area where it would be nice to have fully controlled focus state (related to ideas by @iseulde at #771). So instead of `VisualEditorBlock` finding and calling `focus` on the element directly, the toolbar dispatches the change in focus to the block and the relevant editable detects and calls focus on itself. This is a larger unsolved problem, and not one we really need to account for here.